File: app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/MvvmFragment.java
package openfoodfacts.github.scrachx.openfood.fragments;

import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

import io.reactivex.disposables.CompositeDisposable;
import openfoodfacts.github.scrachx.openfood.views.viewmodel.ViewModel;

public abstract class MvvmFragment<T extends ViewModel, U> extends Fragment {

    private U component;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    @CallSuper
    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        inject();
    }

    @CallSuper
    @Override
    public void onStart() {
        super.onStart();
        bindViewModel();
        bindProperties(compositeDisposable);
    }

    protected void bindViewModel() {
        getViewModel().bind();
    }

    @CallSuper
    @Override
    public void onStop() {
        super.onStop();
        compositeDisposable.clear();
        getViewModel().unbind();
    }

    @NonNull
    public U component() {
        if (component == null) {
            component = createComponent();
        }
        return component;
    }

    protected abstract T getViewModel();

    @NonNull
    protected abstract U createComponent();

    protected abstract void inject();

    protected abstract void bindProperties(CompositeDisposable compositeDisposable);
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/MvvmFragment.kt
package openfoodfacts.github.scrachx.openfood.features

import android.os.Bundle
import androidx.annotation.CallSuper
import androidx.fragment.app.Fragment
import io.reactivex.disposables.CompositeDisposable
import openfoodfacts.github.scrachx.openfood.features.viewmodel.BaseViewModel

abstract class MvvmFragment<T :  BaseViewModel, U> : Fragment() {
    protected val component: U? by lazy {createComponent()}
    private val compositeDisposable = CompositeDisposable()
    @CallSuper
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        inject()
    }

    @CallSuper
    override fun onStart() {
        super.onStart()
        bindViewModel()
        bindProperties(compositeDisposable)
    }

    private fun bindViewModel() {
        viewModel.bind()
    }

    @CallSuper
    override fun onStop() {
        super.onStop()
        compositeDisposable.clear()
        viewModel.unbind()
    }


    protected abstract val viewModel: T
    protected abstract fun createComponent(): U
    protected abstract fun inject()
    protected abstract fun bindProperties(compositeDisposable: CompositeDisposable?)
}